BlueprintUI/Sources/Element/ElementState.swift

```diff
@@ -0,0 +1,597 @@
+        signpostRef: AnyObject,
+        name: String
+    ) {
+        self.parent = parent
+        self.identifier = identifier
+        self.element = ElementSnapshot(element)
+
+        if element is AnyComparableElement {
+            comparability = .comparableElement
+        } else {
+            if let parent = parent {
+                switch parent.comparability {
+                case .notComparable:
+                    comparability = .notComparable
+                case .comparableElement, .childOfComparableElement:
+                    comparability = .childOfComparableElement
+                }
+            } else {
+                comparability = .notComparable
+            }
+        }
+
+        self.signpostRef = signpostRef
+        self.name = name
+
+        wasVisited = true
+        wasUpdateEquivalent = false
+    }
+
+    /// Assigned once per layout cycle, this value represents the
+    /// `Element.content` value from the represented element.
+    /// This value is cached to improve performance (fewer allocations and
+    /// less repeated content building), as well as improves debuggability.
+    private var cachedContent: ElementContent?
+
+    /// Allows the layout system in `ElementContent` to access the `ElementContent`
+    /// of the element, returning a cached version after the first access during a layout cycle.
+    var elementContent: ElementContent {
+        if let cachedContent = cachedContent {
+            return cachedContent
+        } else {
+            let content = element.value.content
+            cachedContent = content
+            return content
+        }
+    }
+
+    /// Updates the element state with the new element and new environment,
+    /// throwing away cached values if needed.
+    ///
+    /// This method takes into account the `Comparability` of the element.
+    ///
+    /// ## Note
+    /// This method may only be called once per layout cycle.
+    fileprivate func update(
+        with newElement: Element,
+        in newEnvironment: Environment,
+        identifier: ElementIdentifier
+    ) {
+        precondition(wasVisited == false)
+        precondition(self.identifier == identifier)
+        precondition(type(of: newElement) == type(of: element.value))
+
+        let isEquivalent: Bool
+
+        switch comparability {
+        case .notComparable:
+            isEquivalent = false
+
+        case .comparableElement:
+            isEquivalent = Self.elementsEquivalent(
+                element.value, newElement,
+                in: ComparableElementContext(
+                    environment: newEnvironment
+                )
+            )
+
+        case .childOfComparableElement:
+            /// We're always, equivalent, because our parent
+            /// determines our equatability from its state.
+            /// It will also invalidate and throw away _our_ caches if its
+            /// equivalency changes.
+            isEquivalent = true
```

Comment:
Nit/thought: I understand that this is correct according to the logic, but it's a bit misleading in terms of the naming and how it gets set on `wasUpdateEquivalent`. Perhaps `wasUpdateEquivalent` and this local val should be optional or a tri-state enum to represent the fact that no update was made?

Reply:
I'm just going to remove `wasUpdateEquivalent` for now, given it was only added for debugging. We could always bring it back if needed